[RF-DOCS] Active Record Callbacks (#51654)
rails/rails: guides/source/active_record_callbacks.md

Active Record Callbacks
=======================

This guide teaches you how to hook into the life cycle of your Active Record
objects.

After reading this guide, you will know:

* When certain events occur during the life of an Active Record object.
* How to register, run, and skip callbacks that respond to these events.
* How to create relational, association, conditional, and transactional
  callbacks.

> * How to create special classes that encapsulate common behavior for your

Reviewer: Maybe "custom objects" or just "objects"

> * How to create special classes that encapsulate common behavior for your

Author: I'll go with "objects", thanks @carlosantoniodasilva and @MatheusRich!

  callbacks to be reused.

--------------------------------------------------------------------------------

The Object Life Cycle
---------------------

During the normal operation of a Rails application, objects may be [created,
updated, and
destroyed](active_record_basics.html#crud-reading-and-writing-data). Active
Record provides hooks into this object life cycle so that you can control your
application and its data.

Callbacks allow you to trigger logic before or after a change to an
object's state. They are methods that get called at certain moments of an
object's life cycle. With callbacks it is possible to write code that will run
whenever an Active Record object is initialized, created, saved, updated, deleted, validated,
or loaded from the database.

```ruby
class Baby < ApplicationRecord

>   after_create -> { Rails.logger.info("Congratulations!") }

Reviewer 2: ActiveRecord models do have logging facilities, so maybe there is no need to be explicit with `Rails.logger...` and use local logger instead?

```ruby
after_create -> { logger.info("Congratulations!") }
```

Applies to (probably) all other `puts` -> `Rails.logger.info` changes bellow.

Author: @tadas-s, thanks for bringing this up!  However, I think it may be preferable to maintain the explicitness to ensure consistency thus reducing the potential confusion for new developers who are still familiarizing themselves with the basics.

>   after_create -> { Rails.logger.info("Congratulations!") }

Reviewer 3: I love this idea

>   after_create -> { Rails.logger.info("Congratulations!") }

Author: ✅

end
```

```irb

> irb> baby = Baby.create

Reviewer 4: this may be misunderstood, to test we need to interact with rails console instead irb.

Reviewer 5: Yes, it got changed in: https://github.com/rails/rails/pull/51732
As a side note, `Baby.create` is a bit of a weird example.
It is less boring than `Post`, but maybe we can do something like `BirthdayCake.create`?

Author: I've updated it to BirthdayCake.create, I think its a nice and creative example.

Congratulations!
```

As you will see, there are many life cycle events and you can choose to hook
into any of these either before, after, or even around them.

Callback Registration
---------------------

To use the available callbacks, you need to implement and register them.
Implementation can be done in a multitude of ways like using ordinary methods,
blocks and procs, or defining custom callback objects using classes. Let's go
through each of these implementation techniques.

You can implement the callbacks as a **macro-style method that calls an ordinary

> method** for registration.

Author: Hmmm, that's a good question! I personally think its apparent from the example.

  before_validation :ensure_username_has_value

  private
    def ensure_username_has_value
      if username.blank?
        self.username = email
      end
    end
end
```

The **macro-style class methods can also receive a block**. Consider using this
style if the code inside your block is so short that it fits in a single line:

```ruby
class User < ApplicationRecord
  validates :username, :email, presence: true

  before_validation do
    self.username = email if username.blank?
  end
end
```

Alternatively, you can **pass a proc to the callback** to be triggered.

```ruby
class User < ApplicationRecord
  validates :username, :email, presence: true

  before_validation ->(user) { user.username = user.email if user.username.blank? }
end
```

Lastly, you can define **a custom callback object**, as shown below. We will
cover these [later in more detail](#callback-objects).

```ruby
class User < ApplicationRecord
  validates :username, :email, presence: true

  before_validation AddUsername
end

class AddUsername
  def self.before_validation(record)
    if record.username.blank?
      record.username = record.email
    end
  end
end
```

> ### Registering Callbacks to Fire on Lifecycle Events

Reviewer 6: Even though the example shows it, should we mention in the text that this is done via the `:on` option? I'm looking to make a connection to the Transaction callbacks where we have an `after_commit on: :destroy` method, and a NOTE afterwards.

> ### Registering Callbacks to Fire on Lifecycle Events

Reviewer 7: What do you think about explaining how to execute save and validate operations with custom contexts? Maybe this should also be included in the validation guides. For example with the `save`, `valid?` methods, etc.

Inspired by [this DHH's gist](https://gist.github.com/dhh/9672827)

Author: I took a pass at this [here](https://github.com/rails/rails/pull/51654/commits/802c601c7b3f531f5286bd0a9620407c3c00594f)

Reviewer 7: Great! I would also add the example with `save`, like: `save(context: :confirmation)`, or whatever.

Do you think it is necessary to add that to the `save` method you can also pass the `touch: false` option? And that would not touch the element, therefore the callbacks would not be executed either. I'm not so sure if that escapes a little of the guide, because in fact when you do that, you are not touching the element, so you are not skipping the callback actually.

Author: > I would also add the example with save, like: save(context: :confirmation), or whatever.

Sounds good! I reused an example to show how you can run validations with a custom context as well as for a methis like `save`.

> Do you think it is necessary to add that to the save method you can also pass the touch: false option? 

Regarding the `touch: false option`, while it's true that it prevents the timestamp update and thus skips the associated callbacks, it's not directly related to validation contexts. Including it in the same discussion might be a bit tangential to the main topic of validation contexts, hence I decided to omit it.

For instance, avoid calling `update(attribute: "value")` inside a callback. This
practice can modify the model's state and potentially lead to unforeseen side
effects during commit. <br><br> Instead, you can assign values directly (e.g.,
`self.attribute = "value"`) in `before_create`, `before_update`, or earlier
callbacks for a safer approach.

#### Custom Contexts

You can define your own custom contexts for callbacks. This can be useful when
you want to perform validations based on specific scenarios, or you want to
group certain callbacks together and run them in a specific context. In these
cases you may be tempted to [skip callbacks](#skipping-callbacks) altogether,
but defining a custom context can sometimes be an alternative structured
approach. You will need to combine a `context` with the `on` option to define a
custom context for a callback.

A common scenario for custom contexts is when you have a multi-step form where
you want to perform validations per step. You can define custom context for each
step of the form:

```ruby
class User < ApplicationRecord
  validate :personal_information, on: :personal_info
  validate :contact_information, on: :contact_info

>   validate :location_information, on: :location_info

Author: Discussion summary with @p8: 

Diving into the code it seems custom contexts are only allowed in `*_validation` callbacks.

For example, `before save` [does not allow](https://github.com/rails/rails/blob/a6a840f150d2d153af33648d93fc4553f5a3857e/activerecord/test/cases/callbacks_test.rb#L476-L483) the `on:` options and *_commit callbacks [only allow](https://github.com/rails/rails/blob/a6a840f150d2d153af33648d93fc4553f5a3857e/activerecord/test/cases/transaction_callbacks_test.rb#L460) `:create`, `:update` and `:destroy`.
So I think it makes sense to move it to the "Active Record Validations" guide and link to it from the "Callbacks" guide.

[`around_create`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-around_create
[`around_save`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-around_save
[`before_create`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-before_create
[`before_save`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-before_save
[`before_validation`]:
    https://api.rubyonrails.org/classes/ActiveModel/Validations/Callbacks/ClassMethods.html#method-i-before_validation

There are examples below that show how to use these callbacks. We've grouped
them by the operation they are associated with, and lastly show how they can
be used in combination. `after_commit` / `after_rollback` examples can be found
[here](active_record_callbacks.html#after-commit-and-after-rollback).

#### Validation Callbacks

Validation callbacks are triggered whenever the record is validated directly via
the
[`valid?`](https://api.rubyonrails.org/classes/ActiveModel/Validations.html#method-i-valid-3F)
or

> [`validate`](https://api.rubyonrails.org/classes/ActiveModel/Validations.html#method-i-valid-3F)

Reviewer 8: ```suggestion
[`validate`](https://api.rubyonrails.org/classes/ActiveModel/Validations.html#method-i-validate)
```

methods, or indirectly via `create`, `update`, or `save`. They are called before
and after the validation phase.

> NOTE: `validate` is an alias for [`valid?`](https://api.rubyonrails.org/classes/ActiveRecord/Validations.html#method-i-valid-3F).

Reviewer: Shoudl we mention `invalid?` too?

> NOTE: `validate` is an alias for [`valid?`](https://api.rubyonrails.org/classes/ActiveRecord/Validations.html#method-i-valid-3F).

Author: Sounds good, thanks @carlosantoniodasilva. Updated to add `invalid?` too @MatheusRich, thanks!

```ruby
class User < ApplicationRecord
  validates :name, presence: true
  before_validation :titleize_name
  after_validation :check_errors

  private
    def titleize_name
      self.name = name.downcase.titleize if name.present?
      Rails.logger.info("Name titleized to #{name}")
    end

    def check_errors
      if errors.any?
        Rails.logger.error("Validation failed: #{errors.full_messages.join(', ')}")
      end
    end
end
```

```irb
irb> user = User.new(name: "", email: "john.doe@example.com", password: "abc123456")
=> #<User id: nil, email: "john.doe@example.com", created_at: nil, updated_at: nil, name: "">

irb> user.valid?
Name titleized to
Validation failed: Name can't be blank
=> false
```

#### Save Callbacks

Save callbacks are triggered whenever the record is persisted (i.e. "saved") to
the underlying database, via the `create`, `update`, or `save` methods. They are
called before, after, and around the object is saved.

```ruby
class User < ApplicationRecord
  before_save :encrypt_password
  around_save :log_saving
  after_save :update_cache

  private
    def encrypt_password

>       self.password = BCrypt::Password.create(password)

Author: No, I think this is okay, I'll update accordingly!

      Rails.logger.info("Password encrypted for user with email: #{email}")
    end

    def log_saving
      Rails.logger.info("Saving user with email: #{email}")
      yield
      Rails.logger.info("User saved with email: #{email}")
    end

    def update_cache
      Rails.cache.write("user_data", attributes)
      Rails.logger.info("Update Cache")
    end
end
```

```irb

> irb> user = User.create(name: "Jane Doe", email: "jane.doe@example.com")

Author: oops... ✅

updated, you may want to notify an admin if critical information is changed. In
this case, you can use `after_create` and `after_update` callbacks together.

```ruby
class User < ApplicationRecord
  after_create :send_confirmation_email
  after_update :notify_admin_if_critical_info_updated

  private
    def generate_confirmation_token
      self.confirmation_token = SecureRandom.hex(10)
      Rails.logger.info("Confirmation token generated for: #{email}")
    end

    def send_confirmation_email
      UserMailer.confirmation_email(self).deliver_later
      Rails.logger.info("Confirmation email sent to: #{email}")
    end

    def notify_admin_if_critical_info_updated
      if saved_change_to_email? || saved_change_to_phone_number?
        AdminMailer.user_critical_info_updated(self).deliver_later
        Rails.logger.info("Notification sent to admin about critical info update for: #{email}")
      end
    end
end
```

```irb
irb> user = User.create(name: "John Doe", email: "john.doe@example.com")
Confirmation email sent to: john.doe@example.com
=> #<User id: 1, email: "john.doe@example.com", ...>

irb> user.update(email: "john.doe.new@example.com")

> Notification sent to admin about critical info update for: john.doe@example.com

Reviewer 8: I expect the updated value to be output to the log.
```suggestion
Notification sent to admin about critical info update for: john.doe.new@example.com
```

Author: NIce catch! Thank you!

    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-before_update

WARNING. The `after_save` callback is triggered on both a create and update.
However, it consistently executes after the more specific callbacks
`after_create` and `after_update`, regardless of the sequence in which the macro
calls were made.

We've already covered [validation](#validation-callbacks) and
[save](#save-callbacks) callbacks. `after_commit` / `after_rollback` examples
can be found [here](#after-commit-and-after-rollback).

#### Update Callbacks

```ruby
class User < ApplicationRecord
  before_update :check_role_change
  around_update :log_updating
  after_update :send_update_email

  private
    def check_role_change
      if role_changed?
        Rails.logger.info("User role changed to #{role}")
      end
    end

    def log_updating
      Rails.logger.info("Updating user with email: #{email}")
      yield
      Rails.logger.info("User updated with email: #{email}")
    end

    def send_update_email
      UserMailer.update_email(self).deliver_later
      Rails.logger.info("Update email sent to: #{email}")
    end
end
```

```irb
irb> user = User.find(1)

> => #<User id: 1, email: "john.doe@example.com", created_at: "2024-03-20 16:19:52.405195000 +0000", updated_at: "2024-03-20 16:19:52.405195000 +0000", name: "John Doe", role: "admin" >

Reviewer 8: For role_changed? to be triggered, the value needs to be updated and changed. For that purpose, I'll set the default to user.

```suggestion
=> #<User id: 1, email: "john.doe@example.com", created_at: "2024-03-20 16:19:52.405195000 +0000", updated_at: "2024-03-20 16:19:52.405195000 +0000", name: "John Doe", role: "user" >
```

* [`after_destroy`][]
* [`after_commit`][] / [`after_rollback`][]

[`after_destroy`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-after_destroy
[`around_destroy`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-around_destroy
[`before_destroy`]:
    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-before_destroy

WARNING: `before_destroy` callbacks should be placed before `dependent: :destroy`
associations (or use the `prepend: true` option), to ensure they execute before

> the records are deleted by `dependent: :destroy`.

Author: I moved it to relational callbacks and updated the wording bit

```ruby
class User < ApplicationRecord
  before_destroy :check_admin_count
  around_destroy :log_destroy_operation
  after_destroy :notify_users

  private
    def check_admin_count

>       if User.where(admin: true).count == 1 && admin?

Reviewer 8: This code might be more correct?

```suggestion
      if User.where(role: "admin").count == 1 && admin?
```

  private
    def log_when_books_or_library_touched
      Rails.logger.info("Book/Library was touched")

>     end

Author: Yup!

* `update`
* `update!`
* `valid?`
* `validate`

Additionally, the `after_find` callback is triggered by the following finder
methods:

* `all`
* `first`
* `find`
* `find_by`
* `find_by!`
* `find_by_*`
* `find_by_*!`
* `find_by_sql`
* `last`
* `sole`
* `take`

The `after_initialize` callback is triggered every time a new object of the
class is initialized.

NOTE: The `find_by_*` and `find_by_*!` methods are dynamic finders generated
automatically for every attribute. Learn more about them in the [Dynamic finders
section](active_record_querying.html#dynamic-finders).

> Skipping Callbacks

Reviewer 7: Why are the references to these methods removed here? I think they should be there, they are quite useful.

It would be great if there was an additional example here as in the other sections, calling any of these methods and showing that the callbacks are not executed.

Also, what do you think about adding the supress method? It is quite useful. Just make it clear that it only suppresses the saves. [ActiveRecord::Suppressor](https://api.rubyonrails.org/classes/ActiveRecord/Suppressor.html)

Author: > Why are the references to these methods removed here? 

@brunoprietog what references are you referring to?

> It would be great if there was an additional example here as in the other sections, calling any of these methods and showing that the callbacks are not executed.

We cover the examples in the section above "Available Callbacks", or did you have any specific scenarios that you had in mind that are not covered by the above section?

> Also, what do you think about adding the supress method? It is quite useful. Just make it clear that it only suppresses the saves. [ActiveRecord::Suppressor](https://api.rubyonrails.org/classes/ActiveRecord/Suppressor.html)

I'm not quite sure about this one, @carlosantoniodasilva do you have thoughts on whether it should be added?

Reviewer 7: > @brunoprietog what references are you referring to?

I'm referring to the links of each method that is in the list of methods that ignores callbacks. Now there are no links to the API reference and before there were.

> We cover the examples in the section above "Available Callbacks", or did you have any specific scenarios that you had in mind that are not covered by the above section?

I mean an example calling a method that ignores callbacks and demonstrating that it does not execute them. In the section on skipping callbacks there are no examples and perhaps the existence of one would be useful in my opinion.

Regarding the suppressor, it was a total unknown to me until I read about it in [this article](https://dev.37signals.com/globals-callbacks-and-other-sacrileges/) on the 37signals dev blog. That's why I think it could be useful.

Author: > Regarding the suppressor, it was a total unknown to me until I read about it in [this article](https://dev.37signals.com/globals-callbacks-and-other-sacrileges/) on the 37signals dev blog. That's why I think it could be useful.

I read through the article and the docs and decided to add a section called [Suppressing Callbacks](https://github.com/rails/rails/pull/51654/commits/d9490a0a85bb6e6f06af91cd4d4d1e0fdb89dbe6) which hopefully encapsulates its value while emphasizing cautiousness.

important business rules and application logic in callbacks that you do not want
to bypass. Bypassing them without understanding the potential implications may
lead to invalid data.

Halting Execution
-----------------

As you start registering new callbacks for your models, they will be queued for
execution. This queue will include all of your model's validations, the
registered callbacks, and the database operation to be executed.

The whole callback chain is wrapped in a transaction. If any callback raises an
exception, the execution chain gets halted and a **rollback** is issued. To
intentionally halt a chain use:

```ruby
throw :abort
```

WARNING. Any exception that is not `ActiveRecord::Rollback` or
`ActiveRecord::RecordInvalid` will be re-raised by Rails after the callback
chain is halted. Additionally, it may break code that does not expect methods
like `save` and `update` (which normally try to return `true` or `false`) to
raise an exception.

NOTE: If an `ActiveRecord::RecordNotDestroyed` is raised within `after_destroy`,
`before_destroy` or `around_destroy` callback, it will not be re-raised and the

> `destroy` method will return `false`.

Author: Maybe something like: 

WARNING: Any exception that is not `ActiveRecord::Rollback` or `ActiveRecord::RecordInvalid` will be re-raised by Rails after the callback chain is halted. This behavior applies not only to operations like save and update, which normally return true or false, but also to destroy operations. If an ActiveRecord::RecordNotDestroyed exception is raised within `after_destroy`, `before_destroy`, or `around_destroy` callbacks, it will not be re-raised, and the `destroy` method will return false.

Wanted your thoughts before incorporating it.

callbacks, ensuring thorough documentation and thoughtful testing to mitigate
risks of unintended side effects, performance issues, and test failures.

Halting Execution
-----------------

As you start registering new callbacks for your models, they will be queued for
execution. This queue will include all of your model's validations, the
registered callbacks, and the database operation to be executed.

The whole callback chain is wrapped in a transaction. If any callback raises an
exception, the execution chain gets halted and a **rollback** is issued. To
intentionally halt a chain use:

```ruby
throw :abort
```

WARNING. Any exception that is not `ActiveRecord::Rollback` or
`ActiveRecord::RecordInvalid` will be re-raised by Rails after the callback
chain is halted. Additionally, it may break code that does not expect methods
like `save` and `update` (which normally try to return `true` or `false`) to
raise an exception.

NOTE: If an `ActiveRecord::RecordNotDestroyed` is raised within `after_destroy`,
`before_destroy` or `around_destroy` callback, it will not be re-raised and the
`destroy` method will return `false`.

> Relational Callbacks

Author: I see your point, it is a bit unclear! Reading the descriptions of the Callbacks  - I would have thought that the Relational Callback is actually an Association Callback. However, I worry that changing the name may confuse people because they are already referring to this type of callback as a Relational Callback.

> Relational Callbacks

Reviewer 5: I don't have a strong opinion on this. Googling for "Relational Callbacks" doesn't return much relevant results (besides copies of the guides): a [Rails PR](https://github.com/rails/rails/issues/37273) that's actually about association callbacks 🙃 and a [Dev.to article](https://dev.to/justalever/understanding-active-record-callbacks-2l2g).
I'm not a fan of "Cascading Callbacks" either, so I'm ok with keeping "Relational Callbacks".

That dev.to article does have a nice simple definition of "Relational Callbacks" though:
"Much like Active Record relations, callbacks can be performed when related objects are changed."
Maybe we can change the next line to something like the following?
"Callbacks can be performed when related objects are changed."
Feel free to ignore it though.

> Relational Callbacks

Author: This is a difficult one but they are confusing and should probably be addressed, here's what I did and let me know your thoughts: 

I think both sections are talking about 'associations', but we use 'relations' as sort of a synonym for 'associations' which confuses me a little bit into thinking we may be talking about different base concepts. 

I kept the 'Association Callbacks' title as it is, however, I changed 'Relational Callbacks' to 'Cascading Association Callbacks' and moved the section below `Association Callbacks'. I also changed the definition as per your suggestion and changed terms like 'relationship' to 'association'.

What do you think?

WARNING: Using the Active Record Suppressor, while potentially beneficial for
selectively controlling callback execution, can introduce complexity and
unexpected behavior. Suppressing callbacks can obscure the intended flow of your
application, leading to difficulties in understanding and maintaining the
codebase over time. Carefully consider the implications of suppressing
callbacks, ensuring thorough documentation and thoughtful testing to mitigate
risks of unintended side effects, performance issues, and test failures.

Halting Execution
-----------------

As you start registering new callbacks for your models, they will be queued for
execution. This queue will include all of your model's validations, the
registered callbacks, and the database operation to be executed.

The whole callback chain is wrapped in a transaction. If any callback raises an
exception, the execution chain gets halted and a **rollback** is issued, and the
error will be re-raised.

```ruby
class Product < ActiveRecord::Base
  before_validation do
    raise "Price can't be negative" if total_price < 0
  end
end

Product.create # raises "Price can't be negative"
```

> This unexpectedly breaks code that does not expect methods like `create` and
> `save` to raise exceptions.
> 

Reviewer 5: Can  we move the warning about re-raising here, right after mentioning re-raising?
Or does that break the flow a bit?
Also, I think `ActiveRecord::RecordNotSaved` is still re-raised, so we can remove it here.
```suggestion
This unexpectedly breaks code that does not expect methods like `create` and
`save` to raise exceptions.

NOTE: If an exception occurs during the callback chain, Rails will re-raise
it unless it is an `ActiveRecord::Rollback` or `ActiveRecord::RecordInvalid` exception.
```

Instead, you should use `throw :abort` to intentionally halt the chain. If any
callback throws `:abort`, the process will be aborted and `create` will return
false.

```ruby
class Product < ActiveRecord::Base
  before_validation do
    throw :abort if total_price < 0
  end
end

Product.create # => false
```

> WARNING: If an exception occurs during the callback chain, Rails will re-raise
> it unless it is an `ActiveRecord::Rollback`, `ActiveRecord::RecordNotSaved` or
> `ActiveRecord::RecordInvalid` exception.

Reviewer 5: If we move the WARNING higher up, I think it would be nice to describe `create!` raising as well.
This would make it similar to `destroy!` below.
````suggestion
However, it will raise an `ActiveRecord::RecordNotSaved` when calling
`create!`. This exception indicates that the record was not saved due to
the callback's interruption.

```ruby
User.create! # => raises an ActiveRecord::RecordNotSaved
```
````
```

Author: I like this, thank you!

When `throw :abort` is called in any destroy callback, `destroy` will return
false:

```ruby
class User < ActiveRecord::Base
  before_destroy do
    throw :abort if still_active?
  end
end

User.first.destroy # => false
```

However, it will raise an `ActiveRecord::RecordNotDestroyed` when calling
`destroy!`.

```ruby
User.first.destroy! # => raises an ActiveRecord::RecordNotDestroyed
```

> 
> In addition to the behaviors mentioned, it's important to note that when `throw
> :abort` is called in a `before_* callback` (such as `before_save`,
> `before_create`, or `before_update`), it will raise an
> `ActiveRecord::RecordNotSaved` exception. This exception indicates that the
> record was not saved due to the callback's interruption. Therefore, when using
> `throw :abort` in `before_*` callbacks, you should be prepared to handle the
> `ActiveRecord::RecordNotSaved` exception.

Reviewer 5: With the above mentioned changes, that make clear that ActiveRecord::RecordNotSaved` exception can be raised, I think we can remove this paragraph altogether.
```suggestion
```

Author: ✅

```ruby
class Comment < ApplicationRecord
  before_save :filter_content,
    if: -> { forum.parental_control? },
    unless: -> { author.trusted? }
end
```

The callback only runs when all the `:if` conditions and none of the `:unless`
conditions are evaluated to `true`.

Transaction Callbacks
---------------------

### `after_commit` and `after_rollback`

Two additional callbacks are triggered by the completion of a database
transaction: [`after_commit`][] and [`after_rollback`][]. These callbacks are
very similar to the `after_save` callback except that they don't execute until
after database changes have either been committed or rolled back. They are most
useful when your Active Record models need to interact with external systems
that are not part of the database transaction.

> Consider if the `PictureFile` model, from the previous example, needs to delete

Reviewer 6: Since these have been moved around, there's no "previous" example for the PictureFile, so we'll have to introduce it here.

Author: Good catch! I introduced it and changed the way the Callback objects example works a little bit. Can you please double check that everything looks good?

end
```

NOTE: The `:on` option specifies when a callback will be fired. If you don't

> supply the `:on` option the callback will fire for every action.

Reviewer 6: Maybe instead of "action" we say "live cycle event"?

I was also wondering if we should connect this back to "Registering Callbacks to Fire on Lifecycle Events", where we introduce the `:on` option.

WARNING. When a transaction completes, the `after_commit` or `after_rollback`
callbacks are called for all models created, updated, or destroyed within that
transaction. However, if an exception is raised within one of these callbacks,
the exception will bubble up and any remaining `after_commit` or
`after_rollback` methods will _not_ be executed. As such, if your callback code
could raise an exception, you'll need to rescue it and handle it within the
callback in order to allow other callbacks to run. <br><br> `after_commit` makes
very different guarantees than `after_save`, `after_update`, and
`after_destroy`. For example, if an exception occurs in an `after_save` the
transaction will be rolled back and the data will not be persisted. However,
anything that happens `after_commit` can guarantee the transaction has already
been completed and the data was persisted to the database. More on

> [transactional callbacks](#transaction-callbacks) below. <br><br> In the context

Reviewer 6: This might need some tweaking, as it's moved here with this warning, we're already within the "transaction callbacks" section. Here's a suggestion:

> However, during `after_commit` the data was already persisted to the database, and thus any exception won't roll anything back anymore. 

I think we can also move the last paragraph together here:

>  However, during `after_commit` the data was already persisted to the database, and thus any exception won't roll anything back anymore. Also note that the code executed within `after_commit` or `after_rollback` callbacks is itself not enclosed within a transaction.

Feel free to tweak or rewrite it as necessary.

end
```

WARNING. Using both `after_create_commit` and `after_update_commit` with the
same method name will only allow the last callback defined to take effect, as
they both internally alias to `after_commit` which overrides previously defined

> callbacks with the same method name.

Reviewer 6: The solution for this one is to use `after_save_commit` instead, maybe we should mention it here?

  private
    def log_user_saved_to_db
      Rails.logger.info("User was saved to database")
    end
end
```

```irb
irb> user = User.create # prints nothing

irb> user.save # updating @user
User was saved to database
```

### `after_save_commit`

There is also [`after_save_commit`][], which is an alias for using the
`after_commit` callback for both create and update together:

```ruby
class User < ApplicationRecord
  after_save_commit :log_user_saved_to_db

  private
    def log_user_saved_to_db
      Rails.logger.info("User was saved to database")
    end
end
```

```irb
irb> user = User.create # creating a User
User was saved to database

irb> user.save # updating user
User was saved to database
```

### Transactional Callback Ordering

> By default (from Rails 7.1), callbacks will run in the order they are defined.

Reviewer 6: ```suggestion
By default (from Rails 7.1), transaction callbacks will run in the order they are defined.
```

  private
    def log_user_saved_to_db
      Rails.logger.info("User was saved to database")
    end
end
```

```irb
irb> user = User.create # prints nothing

irb> user.save # updating @user
User was saved to database
```

### `after_save_commit`

There is also [`after_save_commit`][], which is an alias for using the
`after_commit` callback for both create and update together:

```ruby
class User < ApplicationRecord
  after_save_commit :log_user_saved_to_db

  private
    def log_user_saved_to_db
      Rails.logger.info("User was saved to database")
    end
end
```

```irb
irb> user = User.create # creating a User
User was saved to database

irb> user.save # updating user
User was saved to database
```

### Transactional Callback Ordering

By default (from Rails 7.1), transaction callbacks will run in the order they are defined.

However, in prior versions of Rails, when defining multiple transactional
`after_` callbacks (`after_commit`, `after_rollback`, etc), the order in which
the callbacks were run was reversed.

```ruby
class User < ActiveRecord::Base

>   after_commit { Rails.logger.info("this actually gets called second") }

Reviewer 7: This should be the other way around, since it is understood that we are talking about the default behavior of Rails >= 7.1.

Author: Thanks so much! This and the comment below are really good catches!

  after_commit { Rails.logger.info("this actually gets called first") }
end
```

> We can change that reverse behaviour by setting the [following

Reviewer 7: Will this be necessary for the guides? Either way, calling this change should be the opposite. We would have to update the example below in that case as well.

Author: I think it's a good point to note, but I've updated the example, thanks for pointing it out!

supply the `:on` option the callback will fire for every live cycle event. Read
more about `:on` [here](#registering-callbacks-to-fire-on-lifecycle-events.

WARNING. When a transaction completes, the `after_commit` or `after_rollback`
callbacks are called for all models created, updated, or destroyed within that
transaction. However, if an exception is raised within one of these callbacks,
the exception will bubble up and any remaining `after_commit` or
`after_rollback` methods will _not_ be executed. As such, if your callback code
could raise an exception, you'll need to rescue it and handle it within the
callback in order to allow other callbacks to run. <br><br> `after_commit` makes
very different guarantees than `after_save`, `after_update`, and
`after_destroy`. For example, if an exception occurs in an `after_save` the
transaction will be rolled back and the data will not be persisted. However,
during `after_commit` the data was already persisted to the database, and thus
any exception won't roll anything back anymore. Also note that the code executed
within `after_commit` or `after_rollback` callbacks is itself not enclosed

> within a transaction. <br><br> In the context of a single transaction, if you

Author: Thanks @p8 this was SUCJ a good suggestion and makes such and impact to the readability!

> within a transaction. <br><br> In the context of a single transaction, if you

Reviewer 5: Great! ❤️

> WARNING. When a transaction completes, the `after_commit` or `after_rollback`
> callbacks are called for all models created, updated, or destroyed within that
> transaction. However, if an exception is raised within one of these callbacks,
> the exception will bubble up and any remaining `after_commit` or
> `after_rollback` methods will _not_ be executed. As such, if your callback code
> could raise an exception, you'll need to rescue it and handle it within the
> callback in order to allow other callbacks to run. <br><br> `after_commit` makes
> very different guarantees than `after_save`, `after_update`, and
> `after_destroy`. For example, if an exception occurs in an `after_save` the
> transaction will be rolled back and the data will not be persisted. However,
> during `after_commit` the data was already persisted to the database, and thus
> any exception won't roll anything back anymore. Also note that the code executed
> within `after_commit` or `after_rollback` callbacks is itself not enclosed
> within a transaction. <br><br> In the context of a single transaction, if you
> interact with multiple loaded objects that represent the same record in the
> database, there's a crucial behavior in the `after_commit` and `after_rollback`
> callbacks to note. These callbacks are triggered only for the first object of
> the specific record that changes within the transaction. Other loaded objects,
> despite representing the same database record, will not have their respective
> `after_commit` or `after_rollback` callbacks triggered. This nuanced behavior is
> particularly impactful in scenarios where you expect independent callback
> execution for each object associated with the same database record. It can
> influence the flow and predictability of callback sequences, leading to
> potential inconsistencies in application logic following the
> transaction.<br><br>

Reviewer 5: ````suggestion
interact with multiple loaded objects that represent the same record in the
database, there's a crucial behavior in the `after_commit` and `after_rollback`
callbacks to note. These callbacks are triggered only for the first object of
the specific record that changes within the transaction. Other loaded objects,
despite representing the same database record, will not have their respective
`after_commit` or `after_rollback` callbacks triggered.

```ruby
class User < ApplicationRecord
  after_commit :log_user_saved_to_db, on: :update

  private
    def log_user_saved_to_db
       Rails.logger.info("User was saved to database")
    end
end
```

```ruby
irb> user = User.create

irb> User.transaction { user.save; user.save }
# User was saved to database
```

WARNING: This nuanced behavior is
````
